### PR TITLE
Scopes topic guide

### PR DESCRIPTION
**Pull request by erikadoyle**
Here's an initial concepts doc for scopes (Issue #211). Some open questions I have from @wonderflow's [Health Scope Controller](https://github.com/oam-dev/rudr/tree/master/healthscope) walkthrough:

1. I'm getting this error when I try to `helm install healthscope ./charts/healthscope`: **Error: apiVersion "apps/v1beta2" in healthscope/templates/deployment.yaml is not available** . If I change the apiVersion to e.g., `apps/v1` it (I think?) installs fine. I also tried deleting rudr/crds and reinstalling the latest (w/o specifying version tag), but same results.

2. After applying *health-scope-config.yaml*, the output of `kubectl get health` is **No resources found**.  Yet the output of `kubectl get scopes` lists both **health** and **network**, and *my-health-scope* is listed from `kubectl get configurations`.

3. Is network scope actually implemented in Rudr yet? (Should that be documented w/this PR?)

4. What are the best practices we should include for:
    * When to create a new scope
    * How to share one scope across multiple app configs

Thanks!

### CODE REVIEW

**Comment by wonderflow** [comment]
I could give some answers.

> Question 1

Yes, `apps/v1beta2` is better, I'll fix it.

> Question 2

`kubectl get health` will get real health instance existed, if you apply a `health-scope-config.yaml`, you will find `my-health-scope` both listed from `kubectl get configurations` and `kubectl get health`.  I implemented it as [the spec defines](https://github.com/oam-dev/spec/blob/master/6.application_configuration.md#placing-components-into-application-scopes)

> Question 3

No, network scope is not implemented yet. 

> Question 4

I think if user wants to monitor a bunch of component as one, they will hope to create a health scope.

Multiple app configs can share one scope now, just use the same scope instance name.

**Comment by erikadoyle** [comment]
Thanks for your help @wonderflow!

This PR isn't dependent on it, but I'm still having problems with question 2 above (`kubectl get health` returns *No resources found*). I've tried [HealthScope installation](https://github.com/oam-dev/rudr/tree/master/healthscope) several times on both Windows and MacOS with same results. If its useful, here's my output for `kubectl get pods -l "app.kubernetes.io/name=healthscope"`:

NAME | READY | STATUS | RESTARTS | AGE
-|-|-|-|-|
healthscope-5cd44ff89d-mjg7b|1/1|Running|1|18h

But when I try `kubectl get pods -l "app.kubernetes.io/instance=health"`, I get
**No resources found in default namespace.**

**Comment by wonderflow** [comment]
@erikadoyle  Do you forget to apply the `health-scope-config.yaml`?

```
kubectl apply -f examples/health-scope-config.yaml
```

Health scope instance is created by appconfig

**Comment by erikadoyle** [comment]
Yup!

1. `helm install healthscope ./charts/healthscope` returns **STATUS: deployed**
2. `kubectl apply -f examples/health-scope-config.yaml` returns **applicationconfiguration.core.oam.dev/my-health-scope created**
3. `kubectl get health` returns **No resources found in default namespace**

Is there anything else I should check? Maybe I'm missing a step.

**Comment by wonderflow** [comment]
This could be related with #441 , not resources will be created by that issue, could you re-install rudr and have another try.

**Comment by erikadoyle** [comment]
I uninstalled rudr (and crds), pulled down the latest and reinstalled everything, but still getting no resources found with `kubectl get health`. Here is the log from my rudr pod:

```
[2019-11-05T19:01:10Z INFO  rudr] starting server
[2019-11-05T19:01:10Z INFO  rudr] Loading in-cluster config
[2019-11-05T19:01:10Z INFO  rudr] Watcher is running
[2019-11-05T19:01:10Z INFO  rudr] Health server is running on 0.0.0.0:8080
[2019-11-05T19:01:10Z INFO  rudr] Loading in-cluster config
...
[2019-11-05T19:59:30Z INFO  rudr::instigator] UID: fa08081a-8458-4238-9f11-07b8e28dc641
[2019-11-05T19:59:30Z INFO  rudr::instigator] Looking up first-app
[2019-11-05T19:59:30Z INFO  rudr::instigator] Adding component helloworld-python-v1
[2019-11-05T19:59:30Z INFO  rudr::schematic::component] Looking for image pull secret
[2019-11-05T19:59:31Z INFO  rudr] Handled event
[2019-11-05T19:59:33Z INFO  rudr] Loading in-cluster config
...
[2019-11-05T20:11:41Z INFO  rudr::schematic::scopes::health] health scope my-health-scope created
[2019-11-05T20:11:41Z INFO  rudr] Handled event
[2019-11-05T20:11:41Z ERROR rudr] Error processing event: ErrorMessage { msg: "health scope modify not implemented" }
[2019-11-05T20:11:41Z INFO  rudr] Handled event
[2019-11-05T20:11:46Z INFO  rudr] Loading in-cluster config
[2019-11-05T20:11:46Z INFO  rudr::instigator] Looking up first-app
[2019-11-05T20:11:56Z INFO  rudr] Loading in-cluster config
...
```

**Comment by wonderflow** [comment]
I can't reproduce this. Let me list all the commands I did, then we can compare with each other.

## Delete all things:

```
➜ kubectl delete crd -l app.kubernetes.io/part-of=core.oam.dev
customresourcedefinition.apiextensions.k8s.io "applicationconfigurations.core.oam.dev" deleted
customresourcedefinition.apiextensions.k8s.io "componentinstances.core.oam.dev" deleted
customresourcedefinition.apiextensions.k8s.io "componentschematics.core.oam.dev" deleted
customresourcedefinition.apiextensions.k8s.io "healthscopes.core.oam.dev" deleted
customresourcedefinition.apiextensions.k8s.io "scopes.core.oam.dev" deleted
customresourcedefinition.apiextensions.k8s.io "traits.core.oam.dev" deleted
➜  helm list
NAME  	NAMESPACE	REVISION	UPDATED                             	STATUS  	CHART
health	default  	1       	2019-11-04 14:06:17.384802 +0800 CST	deployed	healthscope-0.1.0
rudr  	default  	1       	2019-11-04 15:26:51.767229 +0800 CST	deployed	rudr-0.1.0
➜  helm delete health rudr
release "health" uninstalled
release "rudr" uninstalled
➜   kubectl get cfg
Error from server (NotFound): Unable to list "core.oam.dev/v1alpha1, Resource=applicationconfigurations": the server could not find the requested resource (get applicationconfigurations.core.oam.dev)
➜  kubectl get health
Error from server (NotFound): Unable to list "core.oam.dev/v1alpha1, Resource=healthscopes": the server could not find the requested resource (get healthscopes.core.oam.dev)

## Install rudr and health
➜  helm install rudr ./charts/rudr --set logLevel="rudr=debug"
NAME: rudr
LAST DEPLOYED: 2019-11-06 10:30:56.610899 +0800 CST m=+2.327815248
NAMESPACE: default
STATUS: deployed
NOTES:
Rudr is a Kubernetes controller to manage Configuration CRDs.

It has been successfully installed.
➜   helm install health charts/healthscope
NAME: health
LAST DEPLOYED: 2019-11-06 10:31:11.85955 +0800 CST m=+0.098950748
NAMESPACE: default
STATUS: deployed
NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods -l "app.kubernetes.io/name=healthscope,app.kubernetes.io/instance=health" -o jsonpath="{.items[0].metadata.name}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl port-forward $POD_NAME 8080:80

## Check running:

➜  kubectl get pods
NAME                                  READY   STATUS              RESTARTS   AGE
health-healthscope-54cfbdc67d-47fpk   0/1     ContainerCreating   0          5s
health-healthscope-54cfbdc67d-gkkl4   0/1     Terminating         0          44h
rudr-5d8dcbb487-qm9pz                 0/1     Running             0          20s
rudr-6c7bd57ccd-fc4xg                 0/1     Terminating         0          43h
➜  kubectl get pods
NAME                                  READY   STATUS    RESTARTS   AGE
health-healthscope-54cfbdc67d-47fpk   1/1     Running   0          2m11s
rudr-5d8dcbb487-qm9pz                 1/1     Running   0          2m26s

## Check internel application scopes installed by helm
➜  kubectl get scopes
NAME      AGE
health    2m47s
network   2m47s

## Install health scope instance

➜  kubectl apply -f examples/health-scope-config.yaml
applicationconfiguration.core.oam.dev/my-health-scope created

## Health scope instance has been installed

➜  kubectl get health
NAME              AGE
my-health-scope   2s

```

**Comment by erikadoyle** [comment]
Ahh! Working now for me--when I followed your steps to delete health, I got an error about a corrupted release record and that I had to manually delete the resources. When I did that (and also started a fresh minikube), the remaining steps worked fine and now `kubectl get health`reports *my-health-scope*. Thank you **so much** @wonderflow  for your help troubleshooting this!

**Comment by wonderflow** [comment]
@erikadoyle Good! Feel free to ask !

You can add these steps to some tutorial of health scope, if it is convenient.

File: docs/concepts/scopes.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  components:
+    - name: helloworld-python-v1
+      instanceName: first-app-helloworld-python-v1
+      parameterValues:
+        # ...
+      traits:
+        # ...  
+      applicationScopes:
+        - my-health-scope
+```
```
Comment (wonderflow) [inline docs/concepts/scopes.md]: check the result of health instance created here

Comment (erikadoyle) [reply docs/concepts/scopes.md]: Good call!

**Pull request merged**